spec/unit/ssl/state_machine_spec.rb

```diff
@@ -0,0 +1,133 @@
+require 'spec_helper'
+require 'puppet_spec/files'
+
+require 'puppet/ssl'
+
+describe Puppet::SSL::StateMachine do
+  include PuppetSpec::Files
+
+  let(:machine) { described_class.new }
+  let(:cacert_pem) { File.read(fixtures('unit/ssl/ssl_provider/ca.pem')) }
+  let(:cacert) { OpenSSL::X509::Certificate.new(cacert_pem) }
+  let(:cacerts) { [cacert] }
+
+  let(:crl_pem) { File.read(fixtures('unit/ssl/ssl_provider/crl.pem')) }
+  let(:crl) { OpenSSL::X509::CRL.new(crl_pem) }
+  let(:crls) { [crl] }
+
+  context 'when ensuring CA certs and CRLs' do
+    it 'returns an SSLContext with the loaded CA certs and CRLs' do
+      Puppet::X509::CertProvider.any_instance.stubs(:load_cacerts).returns(cacerts)
+      Puppet::X509::CertProvider.any_instance.stubs(:load_crls).returns(crls)
+
+      ssl_context = machine.ensure_ca_certificates
+
+      expect(ssl_context[:trusted_certs]).to eq(cacerts)
+      expect(ssl_context[:crls]).to eq(crls)
+      expect(ssl_context[:verify_peer]).to eq(true)
+    end
+  end
+
+  context 'NeedCACerts' do
+    let(:state) { Puppet::SSL::StateMachine::NeedCACerts.new }
+    let(:invalid_pem) { "-----BEGIN CERTIFICATE-----\nwhoops\n-----END CERTIFICATE-----" }
```

Comment:
Need to update cert_provider parsing to handle other kinds of invalid PEM, not just invalid pem between BEGIN/END